### PR TITLE
First pass at Wayland -> XWayland copy/paste

### PR DESCRIPTION
**Pull request by wmww**
Still lots of rough edges, but this does work in the simple case. Current limitations:
- No support for XWayland -> Wayland copy/paste yet
- No support for incremental transfers (attempting to copy paste > 65536 bytes will fail)
- Pasting blocks XWayland thread
- Various TODOs that mark things I may not be handling quite right

I intend to address all of these before I consider the feature complete, but I don't see why any of them would stop this being reviewed and merged. I think it's better to merge stuff as I go rather than have one monster PR at the end.

### CODE REVIEW

**Review by RAOF — COMMENTED** [review]
Looks sensible, although I'm not intimately familiar with the X11 clipboard protocols :)

File: src/server/frontend_xwayland/xwayland_clipboard_provider.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+mf::XWaylandClipboardProvider::~XWaylandClipboardProvider()
+{
+    clipboard->unregister_interest(*clipboard_observer);
+    xcb_destroy_window(connection, selection_window);
+    connection.flush();
+}
+
+void mf::XWaylandClipboardProvider::selection_request_event(xcb_selection_request_event_t* event)
+{
+    if (verbose_xwayland_logging_enabled())
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_xwayland/xwayland_clipboard_provider.cpp:129]: Just so you know, one of the constraints is that a lot of the XWayland debugging has a significant performance penalty (rountrips to the X server to get atom names and such) and shouldn't be enabled during production. I believe filtering the normal way would still evaluate the arguments.

File: src/server/frontend_xwayland/xwayland_clipboard_provider.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        XCB_XFIXES_SELECTION_EVENT_MASK_SET_SELECTION_OWNER |
+        XCB_XFIXES_SELECTION_EVENT_MASK_SELECTION_WINDOW_DESTROY |
+        XCB_XFIXES_SELECTION_EVENT_MASK_SELECTION_CLIENT_CLOSE;
+    xcb_xfixes_select_selection_input(connection, selection_window, connection.CLIPBOARD, mask);
+
+    return selection_window;
+}
+
+/// Like read(), but calls read() until the end of the file or the buffer fills up
+auto read_repeatedly(mir::Fd const& fd, uint8_t* buffer, ssize_t size) -> ssize_t
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_xwayland/xwayland_clipboard_provider.cpp]: Fixed

File: src/server/frontend_xwayland/xwayland_clipboard_provider.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    {
+        auto const remaining = size - total_len;
+        if (remaining <= 0)
+        {
+            break; // buffer full
+        }
+        auto const len = read(fd, buffer + total_len, remaining);
+        if (len < 0)
+        {
+            return -1; // error
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_xwayland/xwayland_clipboard_provider.cpp]: Done

**Comment by wmww** [comment]
Perhaps we should close this and review all of #1892 at once

**Comment by wmww** [comment]
Closing in favor of #1892, can re-open if there are objections

**Pull request closed without merging**